### PR TITLE
Bottleneck Layers for Attention MIL and TransMIL models

### PR DESCRIPTION
**Pull request by CostanzaSiniscalchi**
# Description

Added arguments for  FastAI Trainer Config for AttentionMIL and TransMIL models that can add a bottleneck layer of a specified dimension. 
Required dependencies: None
New Args:
        `dim (int)`: dimensionality. Only applicable for TransMIL models. Some performance improvements for various sizes apart from default 512
        `bottleneck (bool)`: Whether to add the bottleneck layer.
       ` bottleneck_dim (int)`: Dimension of bottleneck layer. Defaults to 20

### CODE REVIEW

File: slideflow/mil/_params.py [hunk 1]
```diff
@@ -135,6 +142,10 @@ def __init__(
                 learning rate schedule. Defaults to True.
             epochs (int): Maximum number of epochs. Defaults to 32.
             batch_size (int): Batch size. Defaults to 64.
+            bottleneck (bool): Whether to implement bottleneck layer or not.
```
Comment (matte-esse) [inline slideflow/mil/_params.py:145]: change to: 

Add a bottleneck layer before the output, this particularly useful if one wants to embed slides into a space of an arbitrary size, the embedding can be later saved through method:{<path_to>}.get_last_layer_activation

File: slideflow/mil/_params.py [hunk 2]
```diff
@@ -135,6 +142,10 @@ def __init__(
                 learning rate schedule. Defaults to True.
             epochs (int): Maximum number of epochs. Defaults to 32.
             batch_size (int): Batch size. Defaults to 64.
+            bottleneck (bool): Whether to implement bottleneck layer or not.    
+                Defaults to False.
+            bottleneck_dim (int): dimension of bottleneck layer.
```
Comment (matte-esse) [inline slideflow/mil/_params.py:147]: bottleneck_size

File: slideflow/mil/train/_fastai.py [hunk 3]
```diff
@@ -229,7 +231,25 @@ def _build_fastai_learner(
...
+            n_in, n_out, config.dim, config.bottleneck_dim, config.bottleneck).to(device)
+        log.info(f"Dimension detected: {config.dim}.")
+        if config.bottleneck:
+            log.info(
+                f"Using bottleneck layer with size {config.bottleneck_dim}.")
+    elif config.model_fn.__name__.lower() == 'attention_mil':
+        model = config.model_fn(
+            n_in, n_out, config.bottleneck_dim, config.bottleneck).to(device)
+        if config.dim:
+            log.warn(f"Dimension specified in config is being ignored.")
```
Comment (matte-esse) [inline slideflow/mil/train/_fastai.py:246]: this should only be printed if the user specifies the dimension and doesn't used tranmil

File: slideflow/mil/train/__init__.py [hunk 4]
```diff
@@ -348,7 +356,7 @@ def train_fastai(
     attention_heatmaps: bool = False,
     **heatmap_kwargs
 ) -> None:
-    """Train an aMIL model using FastAI.
+    """Train an a MIL model using FastAI.
```
Comment (matte-esse) [inline slideflow/mil/train/__init__.py:359]: remove 'a'

File: slideflow/mil/train/__init__.py [hunk 5]
```diff
@@ -402,6 +411,7 @@ def train_fastai(
         outcomes,
         bags=all_bags,
         outdir=outdir,
+
```
Comment (matte-esse) [inline slideflow/mil/train/__init__.py:414]: remove line

File: slideflow/mil/models/transmil.py [hunk 6]
```diff
@@ -7,107 +7,120 @@
...
-        self.layer1 = TransLayer(dim=512)
-        self.layer2 = TransLayer(dim=512)
-        self.norm = nn.LayerNorm(512)
-        self._fc2 = nn.Linear(512, self.n_classes)
+        self.layer1 = TransLayer(dim=dim)
+        self.layer2 = TransLayer(dim=dim)
+        self.norm = nn.LayerNorm(dim)
+        self._fc2 = nn.Linear(dim, self.n_classes)
+        self._fc2_bottleneck = nn.Sequential(
+            nn.Linear(dim, bottleneck_dim), nn.Tanh())
```
Comment (matte-esse) [inline slideflow/mil/models/transmil.py:22]: see the perfomances of Tanh vs ReLU on PEOPLE

File: slideflow/mil/models/transmil.py [hunk 7]
```diff
@@ -7,107 +7,120 @@
...
 
         # ---->Translayer x2
-        h = self.layer2(h)  # [B, N, 512]
+        h = self.layer2(h)  # [B, N, dim]
 
         # ---->cls_token
         h = self.norm(h)[:, 0]
 
+        if self.bottleneck:
+            h = self._fc2_bottleneck(h)
```
Comment (matte-esse) [inline slideflow/mil/models/transmil.py:115]: add a `# ---->bottleneck` comment

File: slideflow/mil/models/att_mil.py [hunk 8]
```diff
@@ -26,21 +30,40 @@ def __init__(
...
-        self.encoder = encoder or nn.Sequential(nn.Linear(n_feats, 256), nn.ReLU())
-        self.attention = attention or Attention(256)
-        self.head = head or nn.Sequential(
-            nn.Flatten(), nn.BatchNorm1d(256), nn.Dropout(), nn.Linear(256, n_out)
-        )
+        self.bottleneck = bottleneck
+
+        if self.bottleneck:
+            self.encoder = encoder or nn.Sequential(
+                nn.Linear(n_feats, 256), nn.ReLU())
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:37]: make it so 256 is taken from `dim` argument (test this)

File: slideflow/mil/models/att_mil.py [hunk 9]
```diff
@@ -4,16 +4,20 @@
...
+
     def __init__(
         self,
         n_feats: int,
         n_out: int,
+        bottleneck_dim: int,
+        bottleneck: bool,
         encoder: Optional[nn.Module] = None,
         attention: Optional[nn.Module] = None,
         head: Optional[nn.Module] = None,
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:23]: add documentation for added (and to be added) arguments

File: slideflow/mil/models/att_mil.py [hunk 10]
```diff
@@ -4,16 +4,20 @@
...
 
     Implementation from: https://github.com/KatherLab/marugoto
 
     """
+
     def __init__(
         self,
         n_feats: int,
         n_out: int,
+        bottleneck_dim: int,
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:19]: add documentation for added (and to be added) arguments

File: slideflow/mil/models/att_mil.py [hunk 11]
```diff
@@ -51,10 +74,15 @@ def get_last_layer_activations(self, bags, lens):
...
         embeddings = self.encoder(bags)
-        masked_attention_scores = self._masked_attention_scores(embeddings, lens)
+        masked_attention_scores = self._masked_attention_scores(
+            embeddings, lens)
+
+        weighted_embedding_sums = (
+            masked_attention_scores * embeddings).sum(-2)
+
+        if self.bottleneck:
+            weighted_embedding_sums = self.added_layer(weighted_embedding_sums)
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:84]:
```
if self.bottleneck:
    return self.added_layer(weighted_embedding_sums)
else: 
    return weighted_embedding_sums
```

File: slideflow/mil/models/att_mil.py [hunk 12]
```diff
@@ -26,21 +30,40 @@ def __init__(
...
-        masked_attention_scores = self._masked_attention_scores(embeddings, lens)
+        masked_attention_scores = self._masked_attention_scores(
+            embeddings, lens)
+
+        weighted_embedding_sums = (
+            masked_attention_scores * embeddings).sum(-2)
 
-        weighted_embedding_sums = (masked_attention_scores * embeddings).sum(-2)
+        if self.bottleneck:
+            weighted_embedding_sums = self.added_layer(weighted_embedding_sums)
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:66]: `bottleneck_embedding =`

File: slideflow/mil/_params.py [hunk 13]
```diff
@@ -146,6 +157,9 @@ def __init__(
         self.fit_one_cycle = fit_one_cycle
         self.epochs = epochs
         self.batch_size = batch_size
+        self.bottleneck = bottleneck
+        self.bottleneck_dim = bottleneck_dim
```
Comment (matte-esse) [inline slideflow/mil/_params.py:161]: change to `bottleneck_size`

File: slideflow/mil/models/att_mil.py [hunk 14]
```diff
@@ -4,16 +4,20 @@
...
     Implementation from: https://github.com/KatherLab/marugoto
 
     """
+
     def __init__(
         self,
         n_feats: int,
         n_out: int,
+        bottleneck_dim: int,
+        bottleneck: bool,
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:20]: we can actually have a single argument `bottleneck_size=None` and the bool check can be done on it being or not being None

File: slideflow/mil/models/att_mil.py [hunk 15]
```diff
@@ -26,21 +30,40 @@ def __init__(
...
         super().__init__()
 
-        self.encoder = encoder or nn.Sequential(nn.Linear(n_feats, 256), nn.ReLU())
-        self.attention = attention or Attention(256)
-        self.head = head or nn.Sequential(
-            nn.Flatten(), nn.BatchNorm1d(256), nn.Dropout(), nn.Linear(256, n_out)
-        )
+        self.bottleneck = bottleneck
+
+        if self.bottleneck:
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:35]: much of the `if` content is the same and should be pulled out

File: slideflow/mil/train/_fastai.py [hunk 16]
```diff
@@ -229,7 +231,27 @@ def _build_fastai_learner(
...
     log.info(f"Training model {config.model_fn.__name__} "
              f"(in={n_in}, out={n_out}, loss={config.loss_fn.__name__})")
-    model = config.model_fn(n_in, n_out).to(device)
+
+    if config.model_fn.__name__.lower() == 'transmil' \
+        or config.model_fn.__name__.lower() == 'attention_mil':
+        if config.model_fn.__name__.lower() == 'transmil' and not config.size:
+            config.size = 512
+        elif config.model_fn.__name__.lower() == 'attention_mil' \
+            and not config.size:
```
Comment (matte-esse) [inline slideflow/mil/train/_fastai.py:240]: config.dim

File: slideflow/mil/models/transmil.py [hunk 17]
```diff
@@ -7,107 +7,124 @@
 
 
 class TransMIL(nn.Module):
-    def __init__(self, n_feats: int, n_out: int,):
+    def __init__(self, n_feats: int, n_out: int, size: int, bottleneck_size: int,
```
Comment (matte-esse) [inline slideflow/mil/models/transmil.py:10]: change `size` to `net_dim`

File: slideflow/mil/_params.py [hunk 18]
```diff
@@ -135,6 +141,12 @@ def __init__(
                 learning rate schedule. Defaults to True.
             epochs (int): Maximum number of epochs. Defaults to 32.
             batch_size (int): Batch size. Defaults to 64.
+            bottleneck_size (int): Add a bottleneck layer of a specific size 
+                before the output, this particularly useful if one wants to 
+                embed slides into a space of an arbitrary size, the embedding 
+                can be later saved through method:{}.get_last_layer_activation
+                Defaults to None.
+            size (int): Size of MIL mode. Defaults to None.
```
Comment (matte-esse) [inline slideflow/mil/_params.py:149]: name

File: slideflow/mil/models/transmil.py [hunk 19]
```diff
@@ -7,107 +7,129 @@
...
+        self.norm = nn.LayerNorm(net_dim)
+        self._fc2 = nn.Linear(net_dim, self.n_classes)
+        if self.bottleneck_size:
+            self._fc2_bottleneck = nn.Sequential(
+                nn.Linear(net_dim, bottleneck_size), nn.Tanh())
+            self._fc3 = nn.Linear(bottleneck_size, self.n_classes)
 
     def calculate_attention(self, h):
-        h = self._fc1(h)  # [B, n, 1024] -> [B, n, 512]
+        h = self._fc1(h)  # [B, n, 1024] -> [B, n, dim]
```
Comment (matte-esse) [inline slideflow/mil/models/transmil.py:33]: net_dim

File: slideflow/mil/models/transmil.py [hunk 20]
```diff
@@ -7,107 +7,129 @@
...
+            h = self._fc2_bottleneck(h)
         return h
 
 # -----------------------------------------------------------------------------
 
+
 class TransLayer(nn.Module):
 
-    def __init__(self, norm_layer=nn.LayerNorm, dim=512):
+    def __init__(self, dim, norm_layer=nn.LayerNorm):
```
Comment (matte-esse) [inline slideflow/mil/models/transmil.py:132]: youc can leave 512 as default, put dim back in last spot

File: slideflow/mil/models/transmil.py [hunk 21]
```diff
@@ -141,7 +163,7 @@ def forward(self, x):
 
 
 class PPEG(nn.Module):
-    def __init__(self, dim=512):
+    def __init__(self, dim):
```
Comment (matte-esse) [inline slideflow/mil/models/transmil.py:166]: you can leave 512 as default

File: slideflow/mil/models/clam.py [hunk 22]
```diff
@@ -84,14 +84,16 @@ class _CLAM_Base(nn.Module):
...
+        bottleneck_size: int,
         size: Union[str, List[int]] = "small",
         dropout: bool = False,
         k_sample: int = 8,
         n_classes: int = 2,
         instance_loss_fn: Optional[Callable] = None,
         subtyping: bool = False,
         gate: bool = True,
-        multi_head_attention: bool = False
+        multi_head_attention: bool = False,
```
Comment (matte-esse) [inline slideflow/mil/models/clam.py:95]: remove comma and empty line

File: slideflow/mil/models/clam.py [hunk 23]
```diff
@@ -84,14 +84,16 @@ class _CLAM_Base(nn.Module):
 
     def __init__(
         self,
+        bottleneck_size: int,
```
Comment (matte-esse) [inline slideflow/mil/models/clam.py:87]: new paramters are added to the end, Default should be None right?

File: slideflow/mil/models/clam.py [hunk 24]
```diff
@@ -341,14 +363,18 @@ class CLAM_SB(_CLAM_Base):
...
```
Comment (matte-esse) [inline slideflow/mil/models/clam.py:366]: move to last

File: slideflow/mil/eval.py [hunk 25]
```diff
@@ -371,19 +369,28 @@ def generate_mil_features(
...
+        model = config.model_fn(config.bottleneck_size, size=_size)
         log.info(
             f"Building model {config.model_fn.__name__} (size={_size})")
-    else:
-        model = config.model_fn(n_features, n_out)
+    else: # FastAI
+        if config.net_dim:
+            model= config.model_fn(n_features, n_out, config.bottleneck_size, 
+                net_dim = config.net_dim)
+            log.info(f"Using network dimension {config.net_dim}")
```
Comment (matte-esse) [inline slideflow/mil/eval.py:385]: I would remove these messages 1/2

File: slideflow/mil/eval.py [hunk 26]
```diff
@@ -371,19 +369,28 @@ def generate_mil_features(
...
+        else:
+            model = config.model_fn(n_features, n_out, config.bottleneck_size)
         log.info(f"Building model {config.model_fn.__name__} "
                  f"(in={n_features}, out={n_out})")
+    if config.bottleneck_size:
+        log.info(
+            f"Using bottleneck layer with size {config.bottleneck_size}.")
```
Comment (matte-esse) [inline slideflow/mil/eval.py:392]: 2/2

File: slideflow/mil/train/_fastai.py [hunk 27]
```diff
@@ -157,7 +159,17 @@ def _build_clam_learner(
     model_size = [n_features] + config_size[1:]
     log.info(f"Training model {config.model_fn.__name__} "
              f"(size={model_size}, loss={config.loss_fn.__name__})")
-    model = config.model_fn(size=model_size, n_classes=batch[-1].shape[-1])
+    
+    if config.net_dim:
+        log.warn("net_dim argument passed is being ignored.")
```
Comment (matte-esse) [inline slideflow/mil/train/_fastai.py:164]: I would just throw an error if the user specifies this and a model that can't use it

File: slideflow/mil/models/clam.py [hunk 28]
```diff
@@ -341,14 +363,18 @@ class CLAM_SB(_CLAM_Base):
...
         dropout: bool = False,
         k_sample: int = 8,
         n_classes: int = 2,
         instance_loss_fn: Optional[Callable] = None,
         subtyping: bool = False,
-        gate: bool = True,
+        gate: bool = True
     ) -> None:
+        
+        self.bottleneck_size = bottleneck_size
```
Comment (matte-esse) [inline slideflow/mil/models/clam.py:376]: move after super's init

File: slideflow/mil/models/clam.py [hunk 29]
```diff
@@ -389,13 +420,14 @@ class CLAM_MB(_CLAM_Base):
 
     def __init__(
         self,
+        bottleneck_size: int,
```
Comment (matte-esse) [inline slideflow/mil/models/clam.py:423]: move down

File: slideflow/mil/_params.py [hunk 30]
```diff
@@ -135,6 +141,12 @@ def __init__(
                 learning rate schedule. Defaults to True.
             epochs (int): Maximum number of epochs. Defaults to 32.
             batch_size (int): Batch size. Defaults to 64.
+            bottleneck_size (int): Add a bottleneck layer of a specific size
```
Comment (matte-esse) [inline slideflow/mil/_params.py:144]: invert order to match definition

File: slideflow/mil/_params.py [hunk 31]
```diff
@@ -146,6 +158,8 @@ def __init__(
         self.fit_one_cycle = fit_one_cycle
         self.epochs = epochs
         self.batch_size = batch_size
+        self.bottleneck_size = bottleneck_size
```
Comment (matte-esse) [inline slideflow/mil/_params.py:161]: inver order

File: slideflow/mil/features.py [hunk 32]
```diff
@@ -151,19 +149,28 @@ def _generate_model(
         if isinstance(config, TrainerConfigCLAM):
             config_size = config.model_fn.sizes[config.model_config.model_size]
             _size = [n_features] + config_size[1:]
-            model = config.model_fn(size=_size)
+            model = config.model_fn(bottleneck_size = config.bottleneck_size, size=_size)
```
Comment (matte-esse) [inline slideflow/mil/features.py:152]: spaces ` = `
